File: clients/csharp/Model/Game.cs
namespace SpbAiChamp.Model
{
    /// <summary>
    /// Current game's state
    /// </summary>
    public struct Game
    {
        /// <summary>
        /// Your player's index
        /// </summary>
        public int MyIndex { get; set; }
        /// <summary>
        /// Current tick number
        /// </summary>
        public int CurrentTick { get; set; }
        /// <summary>
        /// Max number of ticks in the game
        /// </summary>
        public int MaxTickCount { get; set; }
        /// <summary>
        /// List of players
        /// </summary>
        public Model.Player[] Players { get; set; }
        /// <summary>
        /// List of planets
        /// </summary>
        public Model.Planet[] Planets { get; set; }
        /// <summary>
        /// List of flying worker groups
        /// </summary>
        public Model.FlyingWorkerGroup[] FlyingWorkerGroups { get; set; }
        /// <summary>
        /// Max number of flying worker groups for one player
        /// </summary>
        public int MaxFlyingWorkerGroups { get; set; }
        /// <summary>
        /// Max distance of direct travel between planets
        /// </summary>
        public int MaxTravelDistance { get; set; }
        /// <summary>
        /// Max number of workers performing building on one planet
        /// </summary>
        public int MaxBuilders { get; set; }
        /// <summary>
        /// Properties of every building type
        /// </summary>
        public System.Collections.Generic.IDictionary<Model.BuildingType, Model.BuildingProperties> BuildingProperties { get; set; }
    
        public Game(int myIndex, int currentTick, int maxTickCount, Model.Player[] players, Model.Planet[] planets, Model.FlyingWorkerGroup[] flyingWorkerGroups, int maxFlyingWorkerGroups, int maxTravelDistance, int maxBuilders, System.Collections.Generic.IDictionary<Model.BuildingType, Model.BuildingProperties> buildingProperties)
        {
            this.MyIndex = myIndex;
            this.CurrentTick = currentTick;
            this.MaxTickCount = maxTickCount;
            this.Players = players;
            this.Planets = planets;
            this.FlyingWorkerGroups = flyingWorkerGroups;
            this.MaxFlyingWorkerGroups = maxFlyingWorkerGroups;
            this.MaxTravelDistance = maxTravelDistance;
            this.MaxBuilders = maxBuilders;
            this.BuildingProperties = buildingProperties;
        }
    
        /// <summary> Read Game from reader </summary>
        public static Game ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new Game();
            result.MyIndex = reader.ReadInt32();
            result.CurrentTick = reader.ReadInt32();
            result.MaxTickCount = reader.ReadInt32();
            result.Players = new Model.Player[reader.ReadInt32()];
            for (int playersIndex = 0; playersIndex < result.Players.Length; playersIndex++)
            {
                result.Players[playersIndex] = Model.Player.ReadFrom(reader);
            }
            result.Planets = new Model.Planet[reader.ReadInt32()];
            for (int planetsIndex = 0; planetsIndex < result.Planets.Length; planetsIndex++)
            {
                result.Planets[planetsIndex] = Model.Planet.ReadFrom(reader);
            }
            result.FlyingWorkerGroups = new Model.FlyingWorkerGroup[reader.ReadInt32()];
            for (int flyingWorkerGroupsIndex = 0; flyingWorkerGroupsIndex < result.FlyingWorkerGroups.Length; flyingWorkerGroupsIndex++)
            {
                result.FlyingWorkerGroups[flyingWorkerGroupsIndex] = Model.FlyingWorkerGroup.ReadFrom(reader);
            }
            result.MaxFlyingWorkerGroups = reader.ReadInt32();
            result.MaxTravelDistance = reader.ReadInt32();
            result.MaxBuilders = reader.ReadInt32();
            int buildingPropertiesSize = reader.ReadInt32();
            result.BuildingProperties = new System.Collections.Generic.Dictionary<Model.BuildingType, Model.BuildingProperties>(buildingPropertiesSize);
            for (int buildingPropertiesIndex = 0; buildingPropertiesIndex < buildingPropertiesSize; buildingPropertiesIndex++)
            {
                Model.BuildingType buildingPropertiesKey;
                Model.BuildingProperties buildingPropertiesValue;
                buildingPropertiesKey = BuildingTypeHelper.ReadFrom(reader);
                buildingPropertiesValue = Model.BuildingProperties.ReadFrom(reader);
                result.BuildingProperties.Add(buildingPropertiesKey, buildingPropertiesValue);
            }
            return result;
        }
    
        /// <summary> Write Game to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(MyIndex);
            writer.Write(CurrentTick);
            writer.Write(MaxTickCount);
            writer.Write(Players.Length);
            foreach (var playersElement in Players)
            {
                playersElement.WriteTo(writer);
            }
            writer.Write(Planets.Length);
            foreach (var planetsElement in Planets)
            {
                planetsElement.WriteTo(writer);
            }
            writer.Write(FlyingWorkerGroups.Length);
            foreach (var flyingWorkerGroupsElement in FlyingWorkerGroups)
            {
                flyingWorkerGroupsElement.WriteTo(writer);
            }
            writer.Write(MaxFlyingWorkerGroups);
            writer.Write(MaxTravelDistance);
            writer.Write(MaxBuilders);
            writer.Write(BuildingProperties.Count);
            foreach (var buildingPropertiesEntry in BuildingProperties)
            {
                var buildingPropertiesKey = buildingPropertiesEntry.Key;
                var buildingPropertiesValue = buildingPropertiesEntry.Value;
                writer.Write((int) (buildingPropertiesKey));
                buildingPropertiesValue.WriteTo(writer);
            }
        }
    
        /// <summary> Get string representation of Game </summary>
        public override string ToString() {
            string stringResult = "Game { ";
            stringResult += "MyIndex: ";
            stringResult += MyIndex.ToString();
            stringResult += ", ";
            stringResult += "CurrentTick: ";
            stringResult += CurrentTick.ToString();
            stringResult += ", ";
            stringResult += "MaxTickCount: ";
            stringResult += MaxTickCount.ToString();
            stringResult += ", ";
            stringResult += "Players: ";
            stringResult += "[ ";
            int playersIndex = 0;
            foreach (var playersElement in Players)
            {
                if (playersIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += playersElement.ToString();
                playersIndex++;
            }
            stringResult += " ]";
            stringResult += ", ";
            stringResult += "Planets: ";
            stringResult += "[ ";
            int planetsIndex = 0;
            foreach (var planetsElement in Planets)
            {
                if (planetsIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += planetsElement.ToString();
                planetsIndex++;
            }
            stringResult += " ]";
            stringResult += ", ";
            stringResult += "FlyingWorkerGroups: ";
            stringResult += "[ ";
            int flyingWorkerGroupsIndex = 0;
            foreach (var flyingWorkerGroupsElement in FlyingWorkerGroups)
            {
                if (flyingWorkerGroupsIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += flyingWorkerGroupsElement.ToString();
                flyingWorkerGroupsIndex++;
            }
            stringResult += " ]";
            stringResult += ", ";
            stringResult += "MaxFlyingWorkerGroups: ";
            stringResult += MaxFlyingWorkerGroups.ToString();
            stringResult += ", ";
            stringResult += "MaxTravelDistance: ";
            stringResult += MaxTravelDistance.ToString();
            stringResult += ", ";
            stringResult += "MaxBuilders: ";
            stringResult += MaxBuilders.ToString();
            stringResult += ", ";
            stringResult += "BuildingProperties: ";
            stringResult += "{ ";
            int buildingPropertiesIndex = 0;
            foreach (var buildingPropertiesEntry in BuildingProperties)
            {
                if (buildingPropertiesIndex != 0) {
                    stringResult += ", ";
                }
                var buildingPropertiesKey = buildingPropertiesEntry.Key;
                stringResult += buildingPropertiesKey.ToString();
                stringResult += ": ";
                var buildingPropertiesValue = buildingPropertiesEntry.Value;
                stringResult += buildingPropertiesValue.ToString();
                buildingPropertiesIndex++;
            }
            stringResult += " }";
            stringResult += " }";
            return stringResult;
        }
    }
}

File: clients/fsharp/Model/Game.fs
#nowarn "0058"

namespace SpbAiChamp
namespace SpbAiChamp.Model

open SpbAiChamp

/// Current game's state
type Game = {
    /// Your player's index
    MyIndex: int;
    /// Current tick number
    CurrentTick: int;
    /// Max number of ticks in the game
    MaxTickCount: int;
    /// List of players
    Players: Model.Player[];
    /// List of planets
    Planets: Model.Planet[];
    /// List of flying worker groups
    FlyingWorkerGroups: Model.FlyingWorkerGroup[];
    /// Max number of flying worker groups for one player
    MaxFlyingWorkerGroups: int;
    /// Max distance of direct travel between planets
    MaxTravelDistance: int;
    /// Max number of workers performing building on one planet
    MaxBuilders: int;
    /// Properties of every building type
    BuildingProperties: Map<Model.BuildingType, Model.BuildingProperties>;
} with

    /// Write Game to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.MyIndex
        writer.Write this.CurrentTick
        writer.Write this.MaxTickCount
        writer.Write this.Players.Length
        this.Players |> Array.iter (fun value ->
            value.writeTo writer )
        writer.Write this.Planets.Length
        this.Planets |> Array.iter (fun value ->
            value.writeTo writer )
        writer.Write this.FlyingWorkerGroups.Length
        this.FlyingWorkerGroups |> Array.iter (fun value ->
            value.writeTo writer )
        writer.Write this.MaxFlyingWorkerGroups
        writer.Write this.MaxTravelDistance
        writer.Write this.MaxBuilders
        writer.Write this.BuildingProperties.Count
        this.BuildingProperties |> Map.iter (fun key value ->
            writer.Write (int key)
            value.writeTo writer )
        ()

    /// Read Game from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        MyIndex = reader.ReadInt32()
        CurrentTick = reader.ReadInt32()
        MaxTickCount = reader.ReadInt32()
        Players = [|for _ in 1 .. reader.ReadInt32() do
                      yield Model.Player.readFrom reader; |]
        Planets = [|for _ in 1 .. reader.ReadInt32() do
                      yield Model.Planet.readFrom reader; |]
        FlyingWorkerGroups = [|for _ in 1 .. reader.ReadInt32() do
                                 yield Model.FlyingWorkerGroup.readFrom reader; |]
        MaxFlyingWorkerGroups = reader.ReadInt32()
        MaxTravelDistance = reader.ReadInt32()
        MaxBuilders = reader.ReadInt32()
        BuildingProperties = [for _ in 1 .. reader.ReadInt32() do
                                 let key = reader.ReadInt32() |> enum
                                 let value = Model.BuildingProperties.readFrom reader;
                                 yield (key, value) ] |> Map.ofList
    }